File: src/server/gameserver/getGdkeyAccounts/models/udkeyPlayerItem.go
package models

import (
	"encoding/binary"
	"encoding/json"

	gameServerModels "github.com/SuperTavor/Puniemu/src/server/gameserver/models"
	userdatamanager "github.com/SuperTavor/Puniemu/src/userDataManager"
)

type UDKeyPlayerItem struct {
	//ID of the player's icon.
	IconID int `json:"iconID"`
	//Name of the player
	PlayerName string `json:"playerName"`
	//The center yokai in the party, which is displayed next to the user's name.
	//Since we don't have this implemented yet, it will default to 2235000.
	PartnerYokaiID int `json:"youkaiId"`
	//Last time the game was played. Since we don't have access to this info just yet, we'll leave this as 1970-01-01 00:00:00
	LastUpdateDate string `json:"lastUpdateDate"`
	//Title of the character (kun, chan, etc...)
	TitleID int    `json:"titleId"`
	GDKey   string `json:"gdkey"`
	//in Puniemu, the UserID is unused - only the udkey and gdkey are used. This will be left as 0.
	UserID string `json:"userId"`
	//First time the game was played.
	StartDate string `json:"playStartDate"`
}

func NewUDKeyPlayerItem(udkey, gdkey string) (*UDKeyPlayerItem, error) {
	var userData gameServerModels.YwpUserData
	userDataJson, err := userdatamanager.GetYwpUser(udkey, gdkey, "ywp_user_data")
	if err != nil {
		return nil, err
	}
	err = json.Unmarshal(userDataJson, &userData)
	if err != nil {
		return nil, err
	}
	playerItem := UDKeyPlayerItem{}

	playerItem.IconID = userData.IconID
	playerItem.PlayerName = userData.PlayerName
	//2235000 until we have access to this property
	playerItem.PartnerYokaiID = 2235000
	startTimestamp, err := userdatamanager.GetYwpUser(udkey, gdkey, "START_DATE")
	if err != nil {
		return nil, err
	}
	startTimeString := userdatamanager.UnixTimeToDate(int64(binary.LittleEndian.Uint64(startTimestamp)))
	playerItem.StartDate = startTimeString
	//Placeholder - this will be changed when implementing login.nhn
	playerItem.LastUpdateDate = "1970-01-01 00:00:00"
	playerItem.TitleID = userData.CharacterTitleID
	playerItem.GDKey = gdkey
	playerItem.UserID = "0"

	return &playerItem, nil
}

File: Src/Server/GameServer/Requests/GetGdkeyAccounts/DataClasses/UdkeyPlayerItem.cs
using Newtonsoft.Json;
using Puniemu.Src.Server.GameServer.DataClasses;

namespace Puniemu.Src.Server.GameServer.Requests.GetGdkeyAccounts.DataClasses
{
    public struct UdkeyPlayerItem
    {
        [JsonProperty("iconId")]
        public PlayerIcon IconID { get; set; } // ID of the player's icon.

        [JsonProperty("playerName")]
        public string PlayerName { get; set; } // Name of the player

        [JsonProperty("youkaiId")]
        public int PartnerYokaiID { get; set; } // The center yokai in the party

        [JsonProperty("lastUpdateDate")]
        public string LastUpdateDate { get; set; } // Last time the game was played

        [JsonProperty("titleId")]
        public PlayerTitle TitleID { get; set; } // Title of the character (kun, chan, etc...)

        [JsonProperty("gdkey")]
        public string GDKey { get; set; } // GDKey

        [JsonProperty("userId")]
        public string UserID { get; set; } // UserID

        [JsonProperty("playStartDate")]
        public string StartDate { get; set; } // First time the game was played

        //because constructors cant be async
        public static async Task<UdkeyPlayerItem?> ConstructAsync(string gdkey)
        {
            YwpUserData userData;
            try
            {
                userData = await UserDataManager.Logic.UserDataManager.GetYwpUserAsync<YwpUserData>(gdkey, "ywp_user_data");
            }
            catch (UserDataManager.Logic.UserDataManager.TableNotFoundException)
            {
                return null;
            }

            UdkeyPlayerItem playerItem = new();

            playerItem.IconID = (PlayerIcon)userData.IconID;
            playerItem.PlayerName = userData.PlayerName;
            var UserDeck = new TableParser.Logic.TableParser((await UserDataManager.Logic.UserDataManager.GetYwpUserAsync<string>(gdkey, "ywp_user_youkai_deck"))!);
            playerItem.PartnerYokaiID = int.Parse(UserDeck.Table[0][1]);
            var startTimestamp = await UserDataManager.Logic.UserDataManager.GetYwpUserAsync<long>(gdkey, "start_date");
            var startTimestampString = DateTimeOffset.FromUnixTimeMilliseconds(startTimestamp).DateTime.ToString("yyyy-MM-dd HH:mm:ss");
            playerItem.StartDate = startTimestampString;
            //Placeholder as this can only be implemented when we're done with login.nhn
            playerItem.LastUpdateDate = await UserDataManager.Logic.UserDataManager.GetYwpUserAsync<string>(gdkey, "lgn_date");
            playerItem.TitleID = (PlayerTitle)userData.CharacterTitleID;
            playerItem.GDKey = gdkey;
            playerItem.UserID = userData.UserID;
            return playerItem;
        }
    }
}
